File: release/npm/lib/extract.js
"use strict";

const fs = require("node:fs");
const fsp = require("node:fs/promises");
const os = require("node:os");
const path = require("node:path");

/**
 * Finds the archive entry whose basename matches `binaryName`. The archive
 * also contains LICENSE/README (goreleaser default), so callers must not
 * assume a fixed path -- search by basename instead.
 *
 * @param {string[]} entryNames - archive-relative paths (may use `/` separators).
 * @param {string} binaryName - e.g. "scythe" or "scythe.exe".
 * @returns {string} the matching entry name.
 * @throws {Error} when no entry matches.
 */
function findBinaryEntry(entryNames, binaryName) {
  const match = entryNames.find((name) => name.split("/").pop() === binaryName);
  if (!match) {
    throw new Error(
      `scythe-cli: could not find "${binaryName}" inside the downloaded archive. ` +
        `Archive contained: ${entryNames.join(", ") || "(empty)"}`,
    );
  }
  return match;
}

/**
 * Extracts the `scythe` binary from a `.tar.gz` archive buffer to `destPath`.
 *
 * @param {Buffer} buffer
 * @param {string} binaryName
 * @param {string} destPath
 */
async function extractTarGz(buffer, binaryName, destPath) {
  const tar = require("tar");
  const { Readable } = require("node:stream");
  const tmpDir = await fsp.mkdtemp(path.join(os.tmpdir(), "scythe-cli-"));
  try {
    const entries = [];
    await new Promise((resolve, reject) => {
      Readable.from(buffer)
        .pipe(tar.x({ cwd: tmpDir, onentry: (entry) => entries.push(entry.path) }))
        .on("finish", resolve)
        .on("error", reject);
    });

    const entryName = findBinaryEntry(entries, binaryName);
    const extractedPath = path.join(tmpDir, entryName);
    await fsp.mkdir(path.dirname(destPath), { recursive: true });
    await fsp.copyFile(extractedPath, destPath);
  } finally {
    await fsp.rm(tmpDir, { recursive: true, force: true });
  }
}

/**
 * Extracts the `scythe.exe` binary from a `.zip` archive buffer to `destPath`.
 *
 * @param {Buffer} buffer
 * @param {string} binaryName
 * @param {string} destPath
 */
async function extractZip(buffer, binaryName, destPath) {
  const yauzl = require("yauzl");

  await fsp.mkdir(path.dirname(destPath), { recursive: true });

  await new Promise((resolve, reject) => {
    yauzl.fromBuffer(buffer, { lazyEntries: true }, (err, zipfile) => {
      if (err) return reject(err);

      let found = false;
      zipfile.readEntry();
      zipfile.on("entry", (entry) => {
        const basename = entry.fileName.split("/").pop();
        if (basename !== binaryName) {
          zipfile.readEntry();
          return;
        }
        found = true;
        zipfile.openReadStream(entry, (streamErr, readStream) => {
          if (streamErr) return reject(streamErr);
          const out = fs.createWriteStream(destPath);
          readStream.pipe(out);
          out.on("finish", () => {
            zipfile.close();
            resolve(undefined);
          });
          out.on("error", reject);
        });
      });
      zipfile.on("end", () => {
        if (!found) {
          reject(new Error(`scythe-cli: could not find "${binaryName}" inside the downloaded zip archive.`));
        }
      });
      zipfile.on("error", reject);
    });
  });
}

module.exports = { findBinaryEntry, extractTarGz, extractZip };

File: release/pypi/src/scythe_sql/extract.py
"""Archive extraction: locate and unpack the scythe binary by basename.

The archive also contains LICENSE/README (goreleaser default), so callers
must not assume a fixed path inside the archive -- search by basename.
"""

from __future__ import annotations

import io
import stat
import tarfile
import zipfile
from pathlib import Path


class BinaryNotFoundError(RuntimeError):
    """Raised when the archive does not contain an entry matching the binary name."""


def find_binary_entry(entry_names: list[str], binary_name: str) -> str:
    """Finds the archive entry whose basename matches `binary_name`."""
    for name in entry_names:
        if name.rstrip("/").split("/")[-1] == binary_name:
            return name
    listing = ", ".join(entry_names) if entry_names else "(empty)"
    raise BinaryNotFoundError(
        f"scythe-sql: could not find '{binary_name}' inside the downloaded archive. Archive contained: {listing}"
    )


def extract_tar_gz(data: bytes, binary_name: str, dest_path: Path) -> None:
    """Extracts the scythe binary from a `.tar.gz` archive to `dest_path`."""
    with tarfile.open(fileobj=io.BytesIO(data), mode="r:gz") as tar:
        names = [member.name for member in tar.getmembers() if member.isfile()]
        entry_name = find_binary_entry(names, binary_name)
        member = tar.getmember(entry_name)
        extracted = tar.extractfile(member)
        if extracted is None:
            raise BinaryNotFoundError(f"scythe-sql: archive entry '{entry_name}' is not a regular file")
        dest_path.parent.mkdir(parents=True, exist_ok=True)
        dest_path.write_bytes(extracted.read())


def extract_zip(data: bytes, binary_name: str, dest_path: Path) -> None:
    """Extracts the scythe binary from a `.zip` archive to `dest_path`."""
    with zipfile.ZipFile(io.BytesIO(data)) as zf:
        names = [info.filename for info in zf.infolist() if not info.is_dir()]
        entry_name = find_binary_entry(names, binary_name)
        dest_path.parent.mkdir(parents=True, exist_ok=True)
        dest_path.write_bytes(zf.read(entry_name))


def make_executable(path: Path) -> None:
    """chmod 0o755 on POSIX; a no-op is not needed on Windows since .exe requires no bit."""
    path.chmod(stat.S_IRWXU | stat.S_IRGRP | stat.S_IXGRP | stat.S_IROTH | stat.S_IXOTH)
